torch/utils/data/_utils/worker.py

```diff
@@ -269,6 +270,9 @@ def _worker_loop(
         from torch.utils.data import IterDataPipe
         from torch.utils.data.graph_settings import apply_random_seed
 
+        # Dataset has been pickled in the main process
+        dataset = pickle.loads(pickled_dataset)
```

Comment:
Is there an opportunity cost to picking large datasets in the main process, compared to worker process?

Reply:
We would want to skip it if we're in a fork context. But other than that, no, because everything **has to be pickled** to be sent to the workers. It's just handled within the multiprocessing Popen [logic](https://github.com/python/cpython/blob/a1c57bcfd2bcbc55ff858407e09c1d8d8cee44e6/Lib/multiprocessing/popen_spawn_posix.py#L47).

That said, I realized this is actually much more complicated than `pickle.dumps`, because the multiprocessing pickler has a special [register](https://github.com/python/cpython/blob/a1c57bcfd2bcbc55ff858407e09c1d8d8cee44e6/Lib/multiprocessing/reduction.py#L46) to handle more challenging types like Locks depending on the context. This was the failing test [here](https://github.com/pytorch/pytorch/blob/main/test/test_dataloader.py#L1065).

I tried for a while to copy the `multiprocessing` logic but didn't find an adequate solution, so removed for now. I still think this would be a nice improvement, because it's a bit silly to pickle the dataset dozens of times. We could do something hacky like try/except with normal pickle, but a better implementation should be possible.